addons/account/models/account_move.py

```diff
@@ -1713,18 +1713,50 @@ def _check_duplicate_supplier_reference(self):
             ))
 
     def _check_balanced(self):
-        ''' Assert the move is fully balanced debit = credit.
-        An error is raised if it's not the case.
-        '''
+        """ Process request to check that moves are balanced: either do it immediately
+            or postpone till the end of transaction for performance sake"""
+
         moves = self.filtered(lambda move: move.line_ids)
         if not moves:
             return
 
+        # TODO in master: flushing should be done only if we need to check balance
+        # immediately, but apparently some other accounting code depends on
+        # this flushing; so always do the flushing to don't break standard or custom modules
+
```

Comment:
The flushing should be done just before running the SQL query, in `_do_check_balanced`

Reply:
Yes, but it doesn't work
https://github.com/odoo/odoo/pull/73895#issuecomment-883419728

Reply:
Your commit https://github.com/odoo/odoo/commit/4b7626250a7c6ab88dde83f0a4a4494e02e3bfae didn't do what I told you to do here.

Put the flush inside of `_do_check_balanced`, not before calling it at one of the two place where it is supposed to be called. (I could not be flushed before doing the precommit)

However, it is true that the tests are still red. This is because of the tests themselves; and they should not rely on a side effect of the constraint.
* `test_reconcile_cash_basis_workflow_multi_currency`: `assertAmountsGroupByAccount` is executing a SQL query without flushing the data, it is a bug that needs to be fixed (even if it is in a test)
* `test_toggle_reconcile` is playing with `invalidate_cache`, it is only working by pure chance. The test should have had `self.env['account.move.line'].flush(['amount_residual'])`

I don't think ignoring bugs by shadowing them is a good idea.

Reply:
If it's just test issues, I'll fix it of course :+1: . I didn't check carefully and thought it's some other code that expect flushing in _check_balanced

Reply:
The proposed solution with `test_toggle_reconcile` doesn't work. I've managed making it work by changing `api.depends` in `_compute_amount_residual`. I don't quite understand how it works before my PR